### ISSUE TITLE
server/bigquery: computed fields ain't working with limit

### ISSUE THREAD

**Issue by rakeshkky** [hasura/graphql-engine#8562]
TL;DR – The `limit` argument for computed fields is not included in final SQL via `LIMIT` clause. BigQuery throws an exception when `LIMIT` is used after `FROM <table-value function>` expression in a computed field `SELECT` sub-query. Navigate to the end of this description to see probable solutions. 

I'll try to elaborate the issue with an example.

Consider the following table-valued function, `fetch_articles`.
```sql
CREATE TABLE FUNCTION `<redacted_project_name>.hasura_test.fetch_articles`(a_id INT64, search STRING) AS (
(
(SELECT t.* FROM hasura_test.article t where t.author_id = a_id AND (t.title like `search` or t.content like `search`))
)
);
```
The above function returns rows of `article` table given author id `a_id` and `search` keyword. A computed field is defined to `author` table with name `search_articles` using the following definition. 
```yaml
    definition:
      function:
        dataset: hasura_test
        name: fetch_articles
      argument_mapping:
        a_id: id
      return_table:
        name: article
        dataset: hasura_test
```
Let's consider the following GraphQL query that includes the computed field, `search_articles`.

```graphql
query {
  hasura_test_author{
    search_articles(args: {search: "%1%"}){
      id
      title
      content
    }
  }
}
```
Generates following SQL query (replaced parameters)
```sql
SELECT
  ARRAY(
    (
      SELECT
        AS STRUCT `fetch_articles1`.`id` AS `id`,
        `fetch_articles1`.`title` AS `title`,
        `fetch_articles1`.`content` AS `content`
      FROM
        `hasura_test`.`fetch_articles`(`a_id` => `t_author1`.`id`, `search` => '%1%') AS `fetch_articles1`
    )
  ) AS `search_articles_2`
FROM
  `hasura_test`.`author` AS `t_author1`
LIMIT
  1000
```
Let's apply `LIMIT` to the computed field selection,
```sql
SELECT
  ARRAY(
    (
      SELECT
        AS STRUCT `fetch_articles1`.`id` AS `id`,
        `fetch_articles1`.`title` AS `title`,
        `fetch_articles1`.`content` AS `content`
      FROM
        `hasura_test`.`fetch_articles`(`a_id` => `t_author1`.`id`, `search` => '%1%') AS `fetch_articles1`
      LIMIT 1
    )
  ) AS `search_articles_2`
FROM
  `hasura_test`.`author` AS `t_author1`
LIMIT
  1000
```
Running the above query in BigQuery cloud console gives the following error.
```
Correlated subqueries that reference other tables are not supported unless they can be de-correlated, such as by transforming them into an efficient JOIN. 
```

We apply the global limit to all `SELECT` queries generated. This results in a query exception for computed field queries. To mitigate this problem, we chose [not to include `LIMIT` clause](https://github.com/hasura/graphql-engine/blob/af6eb1232eefce6d07b644293d3cec5761bef84f/server/src-lib/Hasura/Backends/BigQuery/FromIr.hs#L1306) for computed field sub-queries. It is not an ideal solution (kind of hack), and `limit` argument for computed field queries became obsolete.

I (@rakeshkky) think, we can solve this by the following ways;
1. Investigate further about the error and figure out an alternate SQL that includes `LIMIT` without any exceptions.
2. Don't generate GraphQL schema for `limit` argument for computed fields, and document this limitation at https://hasura.io/docs/latest/graphql/core/index/ in `Databases > BigQuery > Schema > Computed fields`  by updating [this file](https://github.com/hasura/graphql-engine/blob/master/docs/docs/graphql/core/databases/bigquery/schema/computed-fields.mdx).

Try `1` if it's not possible and go with `2`.

**Update:**
Thoughts on this by @soupi 
> say a computed fields returns 1000000 rows, would that use a lot of memory on the data loader side (and the graphql-engine)?

Technically, yes. Probably we should only look for the solution `1`.

**Comment by rakeshkky** [comment]
An interesting observation:
I'm experimenting with [correlated subqueries](https://cloud.google.com/bigquery/docs/reference/standard-sql/subqueries#correlated_subquery_concepts). Consider the following simple SELECT query.
```sql
select id, array(select id from hasura_test.article a where a.author_id = t.id) as article_ids
from hasura_test.author t
```
The above query runs successfully and provides the following results 
```json
{  "id": "1",  "article_ids": ["1"]}
{  "id": "2",  "article_ids": ["3", "2"]}
```
Introducing a `LIMIT` in the subquery
```sql
select id, array(select id from hasura_test.article a where a.author_id = t.id LIMIT 1) as article_ids
from hasura_test.author t
```
raises the following exception

```
Correlated subqueries that reference other tables are not supported unless they can be de-correlated, such as by transforming them into an efficient JOIN. 
```
It is still not known why using `LIMIT` specifically causes the exception.

**Comment by rakeshkky** [comment]
Further, experimenting with SQLs mentioned in the [above comment](https://github.com/hasura/graphql-engine/issues/8562#issuecomment-1150841040), I figured out the following SQL
```sql
select id, array(select id from unnest(array(select id from hasura_test.article a where a.author_id = t.id)) limit 1) as article_ids
from hasura_test.author t
```
which applies `LIMIT` without any exception and gives the following result
```json
{  "id": "1",  "article_ids": ["1"]}
{  "id": "2",  "article_ids": ["2"]}
```
Essentially, we are enumerating results from subquery using `UNNEST` and then applying `LIMIT`. Applying the same for computed field select query,

```sql
SELECT
  ` t_author1 `.` id ` AS ` id `,
  ARRAY (
    SELECT
      AS STRUCT id,
      title,
      content
    FROM
      UNNEST (
        ARRAY(
          (
            SELECT
              AS STRUCT ` fetch_articles1 `.*
            FROM
              ` hasura_test `.` fetch_articles ` (` a_id ` => ` t_author1 `.` id `, ` search ` => '%1%') AS ` fetch_articles1 `
          )
        )
      )
    LIMIT
      1
  )
FROM
  ` hasura_test `.` author ` AS ` t_author1 `
LIMIT
  1000
```
The above SQL executes without any errors and provides expected results.

**Comment by rakeshkky** [comment]
Implementation of the solution mentioned in the above comment is opened in a draft PR at https://github.com/hasura/graphql-engine-mono/pull/4694.

**Comment by sassela** [comment]
@rakeshkky was this addressed by https://github.com/hasura/graphql-engine-mono/pull/4694? Can it be closed now?

**Comment by rakeshkky** [comment]
Closed with https://github.com/hasura/graphql-engine/commit/b4534c1177ee314d458439eb5fc5e9b6326d6ee0
